daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/unparsers/UState.scala

```diff
@@ -344,7 +342,21 @@ abstract class UState(
       // Finished means you won't add data to the end of it any more.
       // It does NOT prevent information like the absoluteBitPos to
       // propagate.
-      dos.setFinished(this)
+      //
+      // Need this clone or we could possibly end up in an infinite loop inside of
+      // DirectOrBuffereDataOutputStream.cleanUp()
+      val finfo = this match {
+        case m: UStateMain => m.cloneForSuspension(dos)
+        case _ => this
+      }
+
+      val newDOS = dos.addBuffered
+      // NOTE: We must call set finished on the original dos with the
+      // UState/finfo before we set this.dataOutputStream to the newly split
+      // DOS, otherwise the call to this.dataOutputStream = newDOS will update the
+      // formatInfo to point to the new DOS when we try to call setFinished.
+      dos.setFinished(finfo)
+      dataOutputStream = newDOS
```

Comment:
I think the order that this happens as mentioned in the comment no longer matters. As long as we call `cloneForSuspension` before we do `dataOutputStream = newDOS` and modify this state this should all work as expected. This is because `setFinished` will use our clone so changing the `dataOutputStream` of this `UState` no longer has any affect like it does now (which was part of the issue). I'm not sure if one order is preferred over the other, but I don't think the comment is needed if we add something similar to the comment mentioned above.